test/e2e/specs/cart.overview.spec.js

```diff
@@ -0,0 +1,74 @@
+import AppHeader from '../page-objects/AppHeader';
+import SwagOverview from '../page-objects/SwagOverview';
+import CartOverview from '../page-objects/CartOverview';
+import CheckoutPersonalInfo from '../page-objects/CheckoutPersonalInfo';
+import {prepareEnvironment} from '../helpers';
+
+describe('Cart', () => {
+    it('should validate that we can continue shopping', () => {
+        prepareEnvironment('/cart.html');
```

Comment:
@wswebcreation this is all very interesting. A different approach than I would take. Here, I would write this:
`CartPage.open()` which would then navigate me to that URL. How do you feel about such an approach

Reply:
This method is preparing the complete set up of the environment, not only opening the page. It handles:

- setting the session for the user
- setting the provided products in the cart
- navigates to the needed page

Imho this should be a helper and not something as part of a PO because imho a PO should say something about the page itself.

If I need to rename it I'm happy to do it

Reply:
@wswebcreation would this be better in a `setup()` ?
Also, I wonder if this would be better if it was more explicit? What is a `prepareEnvironment()`. You told me, but if you didn't, I'd need to look in the method to figure it out. Would it be more helpful if it was just more explicit? For example:
```
AppState.createUserSession()
ShoppingCart.addDummyProducts();
ShoppingCartPage.open();
```

Reply:
I tend to disagree with this. Reason is that I think that handling the storage, and navigating to the page, should not be use in such a way, what if the app changes and you need to use different ways to set up your environment, then you need to change at least 7 files. I would advice for example customers to have this in a helper because it's a workaround, it has nothing to do with the pages itself

Reply:
The setting up for the environment can change, but that doesn't affect the functionality of the app, which is what we are testing in our UI tests. The functionality/behavior. 
I see `prepareEnvironment` falling into 2 categories:
1. You are actually preparing the environment such as configuring databases
2. You are setting the state of the app so that you can interact with the app.
In the current usage of `prepareEnvironment()` it's not clear what it does without looking at the code. Furthermore, what do you do when you have different requirements for different apps? You now need to have multiple `prepareEnvironment()`? And which method does what is always the question?

Rather than defaulting to a  generic method, we can get specific in our Arrange part of the test. Whatever it is that needs to be done to configure the state of the app should be in the Arrange. It should also be clear to the reader about what is being prepped so that we know what we are testing. 

Yes, I agree with you that if the Arrange part of the app changes, then you need to add/remove steps from your tests. But that's a functionality change now that should affect our specifications?

I would come back to you and ask, why don't you just wrap all the common test steps into a method called `runTest()`, then you won't have to worry about any duplication...

Reply:
You want someone who is reading the test to know what the arrangement is without having to dig. You want it to represent the business logic not the implementation. So yeah, 3 lines instead of 1, but those 3 lines shouldn't ever need to change, just perhaps the underlying implementation. It's ok to be a little wordier when writing tests to minimize the context people need to keep in their brains.

Reply:
Hmm, still not convinced, in my opinion this should be kind of like a util/helper where the name should be more descriptive (maybe I need to check that). I would compare this with something like `Login.signIn(data)`. The sign in can have multiple actions like:

- fill username
- fill password
- add birthdate
- accept terms
- click on login button

Then the method is descriptive enough to explain the login process. If you want to know more about what is happening you just click on the method. Same here with the `prepareEnvironment()`.
Secondly I have seen **workaround** to go from page A to F might change in time due to changes in security layers and so on. I don't want to add all that noice to the test, it's not the scope of the test to keep track of that, thats why we have a util/helper

Reply:
`Login.signIn(data)` is great!
Though my code would be slightly more verbose: `LoginPage.login(UserData.valid)`

The issue is that `prepareEnvironment` doesn't tell anyone what has happened. Is it a new user, existing user, 3 items in a cart, cart items empty, etc. Like, there's a good chance if you stepped away for a month and came back, you wouldn't even remember what all the method does and would have to check. Write it so you never have to check. A more obvious name in this context probably means multiple methods.

Not sure what it looks like in JS, but my ideal Java code would be:
```
SauceDemoApp.authenticateUser(UserData.valid());
SauceDemoApp.addItemsToCart(2);
CartPage cartPage = new CartPage();
cartPage.visit();
```
(Where SauceDemoApp ends up being a wrapper for setting/validating state with API calls/cookie adds/local storage funness)

Reply:
Maybe you can join the PR call with @nadvolod and me on Thursday, don't want to make a complete novel here 😉

Reply:
@wswebcreation I'm curious as to where you make the separation here. 
Why not have your tests like this then?
```
it('should do A', () => {
runTest();
assertionA();
}
it('should do B', () => {
runTest();
assertionB();
}
it('should do C', () => {
runTest();
newTestAction();
assertionC();
}
```

Reply:
@wswebcreation I don't actually care what you do so long as it is testing the stuff; I'm just providing feedback based on what I would tell clients for how to make their tests more maintainable. :)